src/jinja2/runtime.py

```diff
@@ -50,19 +48,31 @@ def identity(x):
     return x
 
 
-def markup_join(seq):
-    """Concatenation that escapes if necessary and converts to string."""
+def markup_join(seq, escape_func=html_escape):
+    """
+    Concatenation that escapes if necessary and converts to string.
+
+    .. versionchanged:: 3.0
+        added optional parameter escape_function to make
+        use the contex based escape function
+    """
     buf = []
     iterator = map(soft_str, seq)
     for arg in iterator:
         buf.append(arg)
         if hasattr(arg, "__html__"):
-            return Markup("").join(chain(buf, iterator))
+            return "".join(map(escape_func, chain(buf, iterator)))
```

Comment:
True. But then both the original type annotations and the function description are wrong.
It would return Markup if input has any Markup like element but string if all is string.
Is this intended?
If you say yes then I will change it so it return Markup (and also change the description and type annotation)

Reply:
I know that Markup is a subclass of string so it is formal correct. But as Markup has a different meaning IMO the typing should state this (as this is done in other functions as well).
Looking at the code a bit more in detail I see that functionality is changed even more, as I escaped the strings, which wasn't done before.
I have to replace it. Hope I can do this, this week or next.

Reply:
Just realized that this function is called by no test at all currently.
Only `test_ex.py` `TestExtensions.test_do` is actually calling `CodeGenerator.visit_Concat` that would lead to a call possible call of `runtime.py`s  `markup_join`.

I will add a test especially for the custom escape now but thought it might be good to know that there a number of not catched test cases considering the logic of `visit_Concat`.

This comment is only FYI @davidism